### DISCUSSION TITLE
Implement WCA Scrambles (Plus, add visualisation for non N×N scrambles)

### DISCUSSION THREAD

**Discussion by Alexuty07** [kash/cubedesk#143] [Ideas]
The Scrambles in CubeDesk are often easier than scrambles from other timers, and i realised that this is due to the program generating what are essentially hand-scrambles (a random sequence of moves) instead of the WCA's official method (a random cube state). For big cubes it is fine to use non-official scrambles because they take a long time to generate, but there should be an option to use official scrambles. Also, how come none of the non N×N puzzles have visualisations? Some people have been waiting a loong time for this feature, and it would also make many profiles look better since PBs for non N×Ns look incomplete. How can this be "The most advanced Rubik's Cube timer, trainer, analytics, […]" without having these fundamental features?

**Comment by afedotov** [comment]
This is not true. Actually 3x3 scrambles implemented in CubeDesk using [Scrambow](https://github.com/Owstin/scrambow) library. And they are definitely **random state**. Scrambow's code is based on javascript version of [min2phase](https://github.com/cs0x7f/min2phase) solver created by Chen Shuang. The min2phase is used in many timers, mainly in csTimer, and even in [tnoodle-lib](https://github.com/thewca/tnoodle-lib) used by WCA to generate scrambles. So I wonder why CubeDesk scrambles looks easier to you.

**Reply by JupiLogy** [reply]
3x3 may be random state but 4x4 is definitely not.

**Reply by afedotov** [reply]
Even WCA uses **random move** scrambles for higher order cubes, look at [tnoodle-lib](https://github.com/thewca/tnoodle-lib) code for reference.
Exception is only 4x4 for which is used **random state** and [TPR-4x4x4-Solver](https://github.com/cs0x7f/TPR-4x4x4-Solver) by Chen.
But you right about Scrambow, for now is uses **random moves** even for 4x4. But I'm not sure this may seriously affect cube state complexity order, given that it is 4x4 cube and CubeDesk generating scramble of 46 moves length.

**Reply by JupiLogy** [reply]
Sure, I know how it works.

The following repo has random scrambles for all WCA puzzles and has more community support - lots of known cubers contribute, the repos are more active, etc. I think it should be used so long as the licenses work for this project.

https://github.com/cubing/scramble.cubing.net

**Reply by afedotov** [reply]
Yep, `cubing/scramble` lib from `cubing.js` is a good alternative, at least for current functionality CubeDesk offer. But AFAIK it can provide only common "WCA events scrambles", and does not offer any subsets like ZBLL,ZZLL,LSE and other like that. In the future this might not fit. Ideally will be if Scrambow author or any contributors will incorporate some work done by Chen in csTimer including 4x4 random state solver/scrambler, looking there this is in todo list.

**Comment by vanfigo** [comment]
Hi guys, I know this thread is mainly discussing about scrambles, but what about the preview of the puzzles, I really enjoy using cubedesk, but having no preview it is really questioning myself if is worth it.
